### PR TITLE
BUG: isin casting to float64 for unsigned int and list

### LINKED ISSUE #46485: BUG: isin() give incorrect results for uint64 columns
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
test_df = pd.DataFrame([{'a': 1378774140726870442}], dtype=np.uint64)
test_df[test_df['a'].isin([1378774140726870528])]
                     a
0  1378774140726870442
test_df[test_df['a'].isin([1378774140726870522])]
                     a
0  1378774140726870442
```

### Issue Description

I have column in uin64 and want to check if values in some list. but isin returning incorrect results. I suspect where some Implicit type conversion, cause I was't able to reproduce for other types

### Expected Behavior

For examples I expected empty DataFrame in both cases

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.100-23
Version          : #1 SMP Fri Jan 31 16:16:00 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.1
numpy            : 1.21.2
pytz             : 2020.5
dateutil         : 2.8.2
pip              : 20.0.2
setuptools       : 58.2.0
Cython           : 0.29.24
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : 2.9.1
jinja2           : 3.0.2
IPython          : 7.28.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.4.3
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : 1.4.25
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #46485 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

The argument is cast to int64 which leads to an upcast later on -> to avoid this we can either use object-dtype for int or implement some casting logic which casts values to unsigned if possible

### CODE REVIEW

**Comment by jreback** [comment]
looks fine. cc @jbrockmendel if any comments.

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -446,7 +447,11 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> npt.NDArray[np.bool_]:
         )
 
     if not isinstance(values, (ABCIndex, ABCSeries, ABCExtensionArray, np.ndarray)):
-        values = _ensure_arraylike(list(values))
+        if not is_signed_integer_dtype(comps):
+            # GH#46485 Use object to avoid upcast to float64 later
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:451]: IIUC the problem is that this comes back with int64, then np.find_common_type for int64+uint64 is float64, which is lossy for big integers.

we face a similar problem elsewhere, and ideally id like to re-use some of the logic we use for those.  the place that comes to mind is in Index._find_common_type_compat.

Comment (phofl) [reply pandas/core/algorithms.py:451]: So we should move the logic from ``_find_common_type_compat`` into a helper function we can call in both places?

Comment (jbrockmendel) [reply pandas/core/algorithms.py:451]: if it can be done gracefully thatd be nice.  otherwise i guess a TODO to look into sharing sooner or later

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**